### PR TITLE
Client Server functionality

### PR DESCRIPTION
**Pull request by Sergii-Kozhushko**
Added Client:
- Client makes init request to Server, eg 'HTTP 222'
- understands Server reply 'ERROR <error-message>' and 'LINK <user-link>'
- after init Client can receive user-http-request from server
Added Server:
- makes init handshake with client
- determines if Service port is already takes and send error to Client
- keeps list with Client connections
- throws custom exception NoConnectionToClient
- listens to User requests and sends them to Client
Utils:
- some changes in order not to close stream after read or write operation

### CODE REVIEW

File: client/src/main/java/de/hellfish/ngrok/client/ClientApplication.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+
+@SpringBootApplication
+public class ClientApplication {
+
+    public static void main(String[] args) {
+        SpringApplication.run(ClientApplication.class, args);
+    }
+
```
Comment (elinarite) [inline client/src/main/java/de/hellfish/ngrok/client/ClientApplication.java]: Remove empty line

File: client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.CommandLineRunner;
+import org.springframework.stereotype.Component;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.net.Socket;
+
```
Comment (elinarite) [inline client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java]: Remove empty line

File: client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+@Component
+public class ClientRunner implements CommandLineRunner {
+    private static int servicePort = 2222;
+    private static final String serverHost = "localhost";
+    private static final int serverPort = 8082;
+    private static boolean running = true;
+    private static Socket serverSocket;
+    private static PrintWriter serverOut;
+    private static BufferedReader serverIn;
+
```
Comment (elinarite) [inline client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java]: Remove empty line

File: client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                log.error("Unknown reply from server");
+                return false;
+            }
+        } catch (IOException e) {
+            log.error("Error establishing connection with server", e);
+            return false;
+        }
+        return true;
+    }
+
```
Comment (elinarite) [inline client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java]: Remove empty line

File: server/src/main/java/de/hellfish/ngrok/server/ClientInitRequest.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package de.hellfish.ngrok.server;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
```
Comment (elinarite) [inline server/src/main/java/de/hellfish/ngrok/server/ClientInitRequest.java:5]: Remove empty line

File: server/src/main/java/de/hellfish/ngrok/server/ClientList.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                return cd.getSocket();
+            }
+        }
+        return null;
+    }
+
+    public void add(ClientData cd) {
+        list.add(cd);
+    }
+
```
Comment (elinarite) [inline server/src/main/java/de/hellfish/ngrok/server/ClientList.java]: Remove empty line

File: server/src/main/java/de/hellfish/ngrok/server/ClientData.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    private final int userPort;
+
+    public String getServiceProtocolAndPort() {
+        return userProtocol + " " + userPort;
+    }
+
+    public String getClientHost() {
+        return socket.getLocalAddress() + ":" + socket.getPort();
+    }
+
```
Comment (elinarite) [inline server/src/main/java/de/hellfish/ngrok/server/ClientData.java]: Remove empty line

File: server/src/main/java/de/hellfish/ngrok/server/NoConnectionToClient.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package de.hellfish.ngrok.server;
+
+public class NoConnectionToClient extends RuntimeException {
+    private final String message;
+
+    public NoConnectionToClient(String message) {
+        super(message);
+        this.message = message;
+    }
+
```
Comment (elinarite) [inline server/src/main/java/de/hellfish/ngrok/server/NoConnectionToClient.java]: Remove empty line

File: server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
+package de.hellfish.ngrok.server;
+
+
```
Comment (elinarite) [inline server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java]: Remove empty line

File: server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+@RequiredArgsConstructor
+public class ServerRunner implements CommandLineRunner {
+
+    private static final int clientToServerPort = 8082;
+    private boolean isRunning = true;
+
+    private final ClientList clientConnections;
+    private final Map<Integer, Socket> userConnections = new HashMap<>();
+    ExecutorService executors = Executors.newFixedThreadPool(5);
+
```
Comment (elinarite) [inline server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java]: Remove empty line

File: server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                // TODO add user-link generation
+                String generatedLink = "http://sub1.localhost:9000";
+                ClientData cd = new ClientData(generatedLink, clientSocket,
+                        protocolAndPortOptional.get().getProtocol(), protocolAndPortOptional.get().getPort());
+                clientList.add(cd);
+                log.info("Received proxy request from client " + cd.getClientHost() +": " + protocolAndPort);
+                sendMessageToClient("LINK " + generatedLink, clientSocket);
+            }
+        }
+
```
Comment (elinarite) [inline server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java]: Remove empty line

File: server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            Optional<ClientInitRequest> protocolAndPortOptional = fetchServicePortAndProtocol(clientSocket);
+            if (protocolAndPortOptional.isEmpty()) {
+                log.error("Error on handshake with client");
+                return;
+            }
+            String protocolAndPort = protocolAndPortOptional.get().getValue();
+            // send user-link to client
+            if (clientList.containsProtocolAndPort(protocolAndPort)) {
+                String errorMessage = "ERROR " + protocolAndPort + " is already in use";
+                log.error(errorMessage);
```
Comment (elinarite) [inline server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java]: You don't want to create a separate class where text messages will be stored?

For example:
public static final String MESSAGE = "ERROR %s is already in use";
and then use String.format(MESSAGE, protocolAndPort)

This will make it easier on tests.

File: server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+            } else {
+                ClientInitRequest mp = new ClientInitRequest(Integer.parseInt(port), protocol);
+                return Optional.of(mp);
+            }
+        } catch (IOException e) {
+            log.error("Error reading init request from client", e);
+        }
+        return Optional.empty();
+    }
+
```
Comment (elinarite) [inline server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java]: Remove empty line

File: server/src/main/java/de/hellfish/ngrok/server/UserController.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+        try {
+            HttpRequestSerializer.writeToOutputStream(myHttpRequest, clientSocket.getOutputStream());
+        } catch (IOException e) {
+            log.error("Error sending user request to client", e);
+            return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE).body("Error connecting to ngrok-client");
+        }
+        return ResponseEntity.ok().build();
+    }
+
```
Comment (elinarite) [inline server/src/main/java/de/hellfish/ngrok/server/UserController.java]: Remove empty line

**Review by elinarite — CHANGES_REQUESTED** [review]
The code looks great, but there are a few comments/suggestions.

File: client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java (hunk 2 above)
Comment (Sergii-Kozhushko) [reply client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java]: Between class annotations and import could be 2 empty lines as far as i understand

File: client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java (hunk 4 above)
Comment (Sergii-Kozhushko) [reply client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java]: This is one of variants of codestyle - to leave empty line before the class end curly brace }. Code looks more structural end curly braces of method and method itself are separated this way. So it was done purposely

File: server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java (hunk 12 above)
Comment (Sergii-Kozhushko) [reply server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java]: 1:1 I will do it in next pull requests

File: utils/src/main/java/de/hellfish/ngrok/utils/MyHttpRequest.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+
+import java.util.Map;
+
+@Getter
+@ToString
+@EqualsAndHashCode
+public final class MyHttpRequest {
```
Comment (raymank26) [inline utils/src/main/java/de/hellfish/ngrok/utils/MyHttpRequest.java]: Name?

**Comment by raymank26** [comment]
Tests?

File: server/src/main/java/de/hellfish/ngrok/server/ClientList.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Stores client data: socket connection, link for users and service port.
+ * Example: http://localhost:2222 -> http://sub1.localhost:9000
+ */
+@Getter
+@Component
+public class ClientList {
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/ClientList.java]: Every method should have O(1) complexity

File: server/src/main/java/de/hellfish/ngrok/server/NoConnectionToClient.java [hunk 17]
```diff
@@ -0,0 +1,11 @@
+package de.hellfish.ngrok.server;
+
+public class NoConnectionToClient extends RuntimeException {
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/NoConnectionToClient.java]: `Exception` suffix should be added.

File: server/src/main/java/de/hellfish/ngrok/server/UserController.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    @GetMapping("/**")
+    public ResponseEntity<String> clientRequests(HttpServletRequest request) {
+
+        // http://sub1.localhost:9000/index/api
+        String userURL = request.getScheme() + "://" + request.getServerName() + ":" + request.getServerPort();
+        if (!clientList.containsUserLink(userURL)) {
+            log.warn("Server received wrong link from user: " + userURL);
+            return ResponseEntity.badRequest().body("Link " + userURL +" was not recognized by ngrok-server");
+        }
+        Socket clientSocket = clientList.getSocketByUserLink(userURL);
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/UserController.java]: Just call it once, `containsUserLink` is not needed above.

File: utils/src/main/java/de/hellfish/ngrok/utils/HttpRequestSerializer.java [hunk 19]
```diff
@@ -9,22 +12,29 @@
...
             objectMapper.writeValue(outputStream, request);
         } catch (IOException e) {
             throw new RuntimeException("Can't write http-request to socket: Client is not available", e);
         }
     }
 
-    public static HttpRequest readFromInputStream(InputStream inputStream) {
-        ObjectMapper objectMapper = new ObjectMapper();
+    public static MyHttpRequest readFromInputStream(InputStream inputStream) {
+        JsonFactory jsonFactory = new JsonFactory();
```
Comment (raymank26) [inline utils/src/main/java/de/hellfish/ngrok/utils/HttpRequestSerializer.java]: Don't init JsonFactory in the convertor body method. Configure it once somewhere in constructor method.

File: server/src/main/java/de/hellfish/ngrok/server/UserController.java [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+     * @param request - contains all info about incoming request
+     * @return
+     */
+    @GetMapping("/**")
+    public ResponseEntity<String> clientRequests(HttpServletRequest request) {
+
+        // http://sub1.localhost:9000/index/api
+        String userURL = request.getScheme() + "://" + request.getServerName() + ":" + request.getServerPort();
+        if (!clientList.containsUserLink(userURL)) {
+            log.warn("Server received wrong link from user: " + userURL);
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/UserController.java]: `log.warn("Server received wrong link from user {}", userURL);`

File: utils/src/main/java/de/hellfish/ngrok/utils/HttpRequestConverter.java [hunk 21]
```diff
@@ -26,7 +30,29 @@ public static HttpRequest convert(HttpServletRequest request) {
...
+        URI uri = URI.create(userUrl);
+
+        Map<String, String> headers = new HashMap<>();
+        myHttpRequest.getHeaders().forEach((key, value) -> headers.put(key, value));
+
+        HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
+                .uri(uri)
+                .method(myHttpRequest.getMethod(), HttpRequest.BodyPublishers.ofString(
+                        new String(myHttpRequest.getBody(), StandardCharsets.UTF_8)));
+        for (Map.Entry<String, String> entry : headers.entrySet()) {
```
Comment (raymank26) [inline utils/src/main/java/de/hellfish/ngrok/utils/HttpRequestConverter.java]: It's better to filter headers on line 42.

File: utils/src/main/java/de/hellfish/ngrok/utils/HttpRequestConverter.java [hunk 22]
```diff
@@ -26,7 +30,29 @@ public static HttpRequest convert(HttpServletRequest request) {
...
+            if (entry.getKey().equals("content-length") || entry.getKey().equals("host") ||
```
Comment (raymank26) [inline utils/src/main/java/de/hellfish/ngrok/utils/HttpRequestConverter.java]: Create some Set<String> static field and put the excluded fields.

File: client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+import java.net.Socket;
+
+
+@Slf4j
+@Component
+public class ClientRunner implements CommandLineRunner {
+    private static int servicePort = 2222;
+    private static final String serverHost = "localhost";
+    private static final int serverPort = 8082;
+    private static boolean running = true;
```
Comment (raymank26) [inline client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java]: This variable is not needed. Just throw an exception from `initConnectionWithServer` and on 38 line just put `return`.

File: client/build.gradle [hunk 24]
```diff
@@ -17,7 +17,7 @@ configurations {
 }
 
 dependencies {
-    implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation 'org.springframework.boot:spring-boot-starter:3.1.5'
```
Comment (raymank26) [inline client/build.gradle]: Why?

File: utils/src/main/java/de/hellfish/ngrok/utils/HttpRequest.java
Comment (raymank26) [inline utils/src/main/java/de/hellfish/ngrok/utils/HttpRequest.java:1]: Why this file is deleted?

**Review by raymank26 — CHANGES_REQUESTED** [review]

File: utils/src/main/java/de/hellfish/ngrok/utils/MyHttpRequest.java (hunk 15 above)
Comment (Sergii-Kozhushko) [reply utils/src/main/java/de/hellfish/ngrok/utils/MyHttpRequest.java]: I changed name cause it duplicated one with java.net.http.HttpRequest and they were in the same blocks of code

File: server/src/main/java/de/hellfish/ngrok/server/ClientList.java (hunk 16 above)
Comment (Sergii-Kozhushko) [reply server/src/main/java/de/hellfish/ngrok/server/ClientList.java]: Why? And how can I do it if i must iterate collection?

File: server/src/main/java/de/hellfish/ngrok/server/UserController.java (hunk 18 above)
Comment (Sergii-Kozhushko) [reply server/src/main/java/de/hellfish/ngrok/server/UserController.java]: Sorry, I don't understand.
Call it? what exactly?
Why 'containsUserLink' not needed?

File: client/build.gradle (hunk 24 above)
Comment (Sergii-Kozhushko) [reply client/build.gradle]: Cause I don't use spring boot starter web in Client

File: utils/src/main/java/de/hellfish/ngrok/utils/HttpRequest.java
Comment (Sergii-Kozhushko) [reply utils/src/main/java/de/hellfish/ngrok/utils/HttpRequest.java:1]: I renamed HttpRequest to MyHttpRequest

File: client/build.gradle (hunk 24 above)
Comment (raymank26) [reply client/build.gradle]: You just deleted version tag, didn't you?

File: utils/src/main/java/de/hellfish/ngrok/utils/HttpRequest.java
Comment (raymank26) [reply utils/src/main/java/de/hellfish/ngrok/utils/HttpRequest.java:1]: Leave it as it, I find the new name misleading.

File: utils/src/main/java/de/hellfish/ngrok/utils/MyHttpRequest.java (hunk 15 above)
Comment (raymank26) [reply utils/src/main/java/de/hellfish/ngrok/utils/MyHttpRequest.java]: "My" prefix is meaningless.

File: server/src/main/java/de/hellfish/ngrok/server/ClientList.java (hunk 16 above)
Comment (raymank26) [reply server/src/main/java/de/hellfish/ngrok/server/ClientList.java]: Map<String, Socket>

File: server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class ServerRunner implements CommandLineRunner {
+
+    private static final int clientToServerPort = 8082;
+    private boolean isRunning = true;
+    private final ClientList clientConnections;
+    ExecutorService executors = Executors.newFixedThreadPool(5);
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java]: private? Move a constant to a constructor.

File: server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    }
+
+    private static void sendMessageToClient(String message, Socket clientSocket) {
+        try {
+            PrintWriter pw = new PrintWriter(clientSocket.getOutputStream(), true);
+            pw.write(message + "\n");
+            pw.flush();
+        } catch (IOException e) {
+            throw new NoConnectionToClientException("Error sending service message to client: " + message);
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java]: cause?

**Comment by Sergii-Kozhushko** [comment]
Made all required changes except tests. And also added error messages as constants as Elina suggested.

File: client/build.gradle (hunk 24 above)
Comment (Sergii-Kozhushko) [reply client/build.gradle]: No, I changed spring-boot-starter-web dependency to spring-boot-starter

File: client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+        }
+    }
+
+    private static int fetchServicePort(String[] args) {
+        int servicePort = 0;
+            for (String arg : args) {
+                if (arg.contains("-port=")) {
+                    try {
+                        servicePort = Integer.parseInt(arg.substring(6));
+                    } catch (NumberFormatException ignored) {
```
Comment (raymank26) [inline client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java]: ?

File: client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+            BufferedReader serverIn = new BufferedReader(new InputStreamReader(serverSocket.getInputStream()));
+
+            serverOut.print("HTTP " + servicePort + "\n");
+            serverOut.flush();
+            String serverReply = serverIn.readLine();
+            String serverReplyCode = serverReply.split(" ")[0];
+
+            if (serverReplyCode.equals("LINK")) {
+                String link = serverReply.split(" ")[1];
+                log.info(String.format("Link for user requests: http://localhost:%s -> %s", servicePort, link));
```
Comment (raymank26) [inline client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java]: Don't use `String.format` with `log` together. `Log` is capable for formatting handling as well.

File: config/checkstyle/checkstyle.xml [hunk 29]
```diff
@@ -245,14 +244,14 @@
...
         </module>
-        <module name="Indentation">
-            <property name="basicOffset" value="2"/>
-            <property name="braceAdjustment" value="2"/>
-            <property name="caseIndent" value="2"/>
-            <property name="throwsIndent" value="4"/>
-            <property name="lineWrappingIndentation" value="4"/>
-            <property name="arrayInitIndent" value="2"/>
-        </module>
+<!--        <module name="Indentation">-->
```
Comment (raymank26) [inline config/checkstyle/checkstyle.xml]: Why so?

File: config/checkstyle/checkstyle.xml [hunk 30]
```diff
@@ -263,12 +262,12 @@
         </module>
         <module name="OverloadMethodsDeclarationOrder"/>
         <module name="VariableDeclarationUsageDistance"/>
-        <module name="CustomImportOrder">
-            <property name="sortImportsInGroupAlphabetically" value="true"/>
-            <property name="separateLineBetweenGroups" value="true"/>
-            <property name="customImportOrderRules" value="STATIC###THIRD_PARTY_PACKAGE"/>
-            <property name="tokens" value="IMPORT, STATIC_IMPORT, PACKAGE_DEF"/>
-        </module>
+<!--        <module name="CustomImportOrder">-->
```
Comment (raymank26) [inline config/checkstyle/checkstyle.xml]: Why so? Delete if needed, don't comment.

File: server/src/main/java/de/hellfish/ngrok/server/ServerConfig.java [hunk 31]
```diff
@@ -0,0 +1,17 @@
...
+import org.springframework.context.annotation.Configuration;
+import java.net.Socket;
+import java.util.HashMap;
+import java.util.Map;
+
+@Configuration
+public class ServerConfig {
+
+    @Bean
+    public Map<String, Socket> clientConnections() {
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/ServerConfig.java]: Don't expose `Map` as a Bean.

File: server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java [hunk 32]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class ServerRunner implements CommandLineRunner {
+
+    @Value("${client.port}")
+    private int clientPort;
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java]: final

File: server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java [hunk 33]
```diff
@@ -0,0 +1,123 @@
...
+    private final Map<String, Socket> clientConnections;
+    private final ExecutorService executors = Executors.newFixedThreadPool(5);
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java]: Put it in `application.properties`

File: .gitignore [hunk 34]
```diff
@@ -24,4 +24,6 @@ bin/
 *.ipr
 out/
 !**/src/main/**/out/
-!**/src/test/**/out/
\ No newline at end of file
+!**/src/test/**/out/
+/client/src/test/java/de/hellfish/ngrok/client/ClientRunnerTest.java
```
Comment (raymank26) [inline .gitignore]: Use local git exclude mechanism

File: client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java (hunk 27 above)
Comment (Sergii-Kozhushko) [reply client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java]: Client receives command line argument -port= here. I don't understand what's wrong here

**Comment by Sergii-Kozhushko** [comment]
Exchanged Map<String, Sockets> bean to ClientConnectionService. And now i'am waiting for further review

File: client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java (hunk 27 above)
Comment (raymank26) [reply client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java]: What if I enter "-port=sdlfkjsdf"?

File: server/src/main/java/de/hellfish/ngrok/server/Server.java [hunk 35]
```diff
@@ -0,0 +1,65 @@
...
+@Getter
+@Component
+@RequiredArgsConstructor
+public final class Server implements Runnable {
+    @Value("${client.port}")
+    private int clientPort;
+    private final ClientConnectionService clientConnections;
+    private ServerSocket serverSocket;
+    private boolean running;
+    private final ExecutorService executors = Executors.newFixedThreadPool(5);
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/Server.java]: Use `application.properties` value instead of `5` number.

File: server/src/main/java/de/hellfish/ngrok/server/Server.java [hunk 36]
```diff
@@ -0,0 +1,65 @@
...
+
+    public void startServer() {
+        try {
+            serverSocket = new ServerSocket(clientPort);
+            running = true;
+            log.info("Ngrok-Server started. Listening on clients on port: {}", clientPort);
+        } catch (
+                IOException e) {
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/Server.java]: Formatting.

File: config/checkstyle/checkstyle.xml [hunk 37]
```diff
@@ -245,14 +228,6 @@
             <message key="ws.notPreceded"
                      value="GenericWhitespace ''{0}'' is not preceded with whitespace."/>
         </module>
-        <module name="Indentation">
```
Comment (raymank26) [inline config/checkstyle/checkstyle.xml:248]: Why did you delete it?

File: config/checkstyle/checkstyle.xml [hunk 38]
```diff
@@ -327,33 +295,12 @@
...
-                      RECORD_DEF, ANNOTATION_DEF"/>
-            <property name="excludeScope" value="nothing"/>
-        </module>
         <module name="MethodName">
             <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9_]*$"/>
             <message key="name.invalidPattern"
                      value="Method name ''{0}'' must match pattern ''{1}''."/>
         </module>
         <module name="SingleLineJavadoc"/>
-        <module name="EmptyCatchBlock">
```
Comment (raymank26) [inline config/checkstyle/checkstyle.xml:350]: Don't delete it.

File: config/checkstyle/checkstyle.xml [hunk 39]
```diff
@@ -327,33 +295,12 @@
...
-            <property name="exceptionVariableName" value="expected"/>
-        </module>
-        <module name="CommentsIndentation">
```
Comment (raymank26) [inline config/checkstyle/checkstyle.xml:353]: Why did you delete it?

File: server/src/main/java/de/hellfish/ngrok/server/ClientConnectionService.java [hunk 40]
```diff
@@ -0,0 +1,13 @@
+package de.hellfish.ngrok.server;
+
+import java.net.Socket;
+
+public interface ClientConnectionService {
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/ClientConnectionService.java]: Why an interface is created?

File: server/src/main/java/de/hellfish/ngrok/server/ClientConnectionServiceImpl.java [hunk 41]
```diff
@@ -0,0 +1,31 @@
+package de.hellfish.ngrok.server;
+
+import org.springframework.stereotype.Service;
+
+import java.net.Socket;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Service
+public class ClientConnectionServiceImpl implements ClientConnectionService {
+    private final ConcurrentHashMap<String, Socket> clientConnections = new ConcurrentHashMap<>();
```
Comment (raymank26) [inline server/src/main/java/de/hellfish/ngrok/server/ClientConnectionServiceImpl.java]: empty line between class declaration and a field

File: client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java (hunk 27 above)
Comment (Sergii-Kozhushko) [reply client/src/main/java/de/hellfish/ngrok/client/ClientRunner.java]: I just checked: if i provide -port=sdlfkjsdf, client logs error and finishes. So it is ok with it

File: config/checkstyle/checkstyle.xml (hunk 29 above)
Comment (Sergii-Kozhushko) [reply config/checkstyle/checkstyle.xml]: It always made warnings so i switched it off. But today I managed to set correct Indentation numbers, so i brought it back.

File: server/src/main/java/de/hellfish/ngrok/server/Server.java (hunk 35 above)
Comment (Sergii-Kozhushko) [reply server/src/main/java/de/hellfish/ngrok/server/Server.java]: done!

File: config/checkstyle/checkstyle.xml (hunk 37 above)
Comment (Sergii-Kozhushko) [reply config/checkstyle/checkstyle.xml:248]: I replied upper.

File: config/checkstyle/checkstyle.xml (hunk 39 above)
Comment (Sergii-Kozhushko) [reply config/checkstyle/checkstyle.xml:353]: It didn't work correct way. I fixed it and returned EmptyCatchBlock to checkstyle file.

File: server/src/main/java/de/hellfish/ngrok/server/ClientConnectionServiceImpl.java (hunk 41 above)
Comment (Sergii-Kozhushko) [reply server/src/main/java/de/hellfish/ngrok/server/ClientConnectionServiceImpl.java:10]: To make realization and definition separate. Code looks better, easier to support. App is more flexible and finally support better polymorphism

File: server/src/main/java/de/hellfish/ngrok/server/ClientConnectionService.java (hunk 40 above)
Comment (Sergii-Kozhushko) [reply server/src/main/java/de/hellfish/ngrok/server/ClientConnectionService.java]: To make realization and definition separate. Code looks better, easier to support. App is more flexible and finally support better polymorphism

File: config/checkstyle/checkstyle.xml (hunk 38 above)
Comment (Sergii-Kozhushko) [reply config/checkstyle/checkstyle.xml:350]: Ok, I put it back.

File: server/src/main/java/de/hellfish/ngrok/server/ClientConnectionServiceImpl.java (hunk 41 above)
Comment (Sergii-Kozhushko) [reply server/src/main/java/de/hellfish/ngrok/server/ClientConnectionServiceImpl.java]: Solved

File: server/src/main/java/de/hellfish/ngrok/server/ClientConnectionService.java (hunk 40 above)
Comment (Sergii-Kozhushko) [reply server/src/main/java/de/hellfish/ngrok/server/ClientConnectionService.java]: But finally i red this article 

https://medium.com/w-logs/think-twice-before-implementing-interface-classes-in-microservices-abc67ae11fac

and i should of course remove interface here.

**Comment by Sergii-Kozhushko** [comment]
Redesigned client structure and form of messages, added client test.

File: client/src/main/java/de/hellfish/ngrok/client/NgrokClient.java [hunk 42]
```diff
@@ -0,0 +1,132 @@
...
+                    running = false;
+                }
+            } catch (IOException e) {
+                log.error("Error connecting to server", e);
+                running = false;
+            }
+        }
+        stop();
+    }
+
```
Comment (elinarite) [inline client/src/main/java/de/hellfish/ngrok/client/NgrokClient.java:73]: Take a look, it might be possible to shorten the record by removing one extra try block, leaving one try and two catch blocks

File: client/src/test/java/de/hellfish/ngrok/client/NgrokClientTest.java [hunk 43]
```diff
@@ -0,0 +1,69 @@
...
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }).start();
+        Thread.sleep(1000);
+
```
Comment (elinarite) [inline client/src/test/java/de/hellfish/ngrok/client/NgrokClientTest.java:52]: Take a look, it might be possible to shorten the record by removing one extra try block, leaving one try and two catch blocks

**Review by elinarite — COMMENTED** [review]
Beautiful code, well done :)

File: client/src/main/java/de/hellfish/ngrok/client/NgrokClient.java (hunk 42 above)
Comment (Sergii-Kozhushko) [reply client/src/main/java/de/hellfish/ngrok/client/NgrokClient.java:73]: Fixed!

File: client/src/test/java/de/hellfish/ngrok/client/NgrokClientTest.java (hunk 43 above)
Comment (Sergii-Kozhushko) [reply client/src/test/java/de/hellfish/ngrok/client/NgrokClientTest.java:52]: Also done!

File: client/src/test/java/de/hellfish/ngrok/client/NgrokClientTest.java [hunk 44]
```diff
@@ -0,0 +1,69 @@
...
+                     PrintWriter out = new PrintWriter(socket.getOutputStream(), true)) {
+                    out.println("LINK " + testLink);
+                    try {
+                        Thread.sleep(2000);
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (raymank26) [inline client/src/test/java/de/hellfish/ngrok/client/NgrokClientTest.java]: Never do " e.printStackTrace();" in real life code. For now, you throw "message" from the "e" exception away. Prefer logging setup (slf4j).

File: client/src/test/java/de/hellfish/ngrok/client/NgrokClientTest.java [hunk 45]
```diff
@@ -0,0 +1,69 @@
...
+            }
+        }).start();
+        Thread.sleep(1000);
+
+        client = new NgrokClient(testPort, "http");
+    }
+
+    @Test
+    public void testClientServerInteraction() throws InterruptedException {
+        Thread clientThread = new Thread(client);
```
Comment (raymank26) [inline client/src/test/java/de/hellfish/ngrok/client/NgrokClientTest.java:58]: Give a name to the thread.

File: client/src/main/java/de/hellfish/ngrok/client/NgrokClient.java [hunk 46]
```diff
@@ -0,0 +1,132 @@
...
+    public static void main(String[] args) {
+        if (args.length < 2) {
+            log.error("Provide service protocol and port. Syntax: java NgrokClient <protocol> <port>");
+            return;
+        }
+        String serviceProtocol = args[0];
+        int servicePort = 0;
+        try {
+            servicePort = Integer.parseInt(args[1]);
+        } catch (NumberFormatException ignored) {
```
Comment (raymank26) [inline client/src/main/java/de/hellfish/ngrok/client/NgrokClient.java:45]: Never ignore exceptions on user input.

Comment (Sergii-Kozhushko) [reply client/src/main/java/de/hellfish/ngrok/client/NgrokClient.java:45]: Fixed

File: client/src/test/java/de/hellfish/ngrok/client/NgrokClientTest.java (hunk 44 above)
Comment (Sergii-Kozhushko) [reply client/src/test/java/de/hellfish/ngrok/client/NgrokClientTest.java]: Fixed
